src/main/java/software/amazon/ionpathextraction/SearchPath.java

```diff
@@ -13,21 +13,64 @@
 
 package software.amazon.ionpathextraction;
 
+import java.util.List;
 import java.util.function.BiFunction;
 import software.amazon.ion.IonReader;
+import software.amazon.ionpathextraction.internal.ArrayUtils;
+import software.amazon.ionpathextraction.internal.MatchContext;
+import software.amazon.ionpathextraction.pathcomponents.PathComponent;
 
 /**
  * A path which is provided to the extractor for matching.
  *
  * @param <T> type accepted by the callback function
  */
-interface SearchPath<T> {
+final class SearchPath<T> {
 
-    BiFunction<IonReader, T, Integer> getCallback();
+    private final List<PathComponent> pathComponents;
+    private final BiFunction<IonReader, T, Integer> callback;
+    private final String[] annotations;
 
-    Type getType();
+    SearchPath(final List<PathComponent> pathComponents,
+               final BiFunction<IonReader, T, Integer> callback,
+               final String[] annotations) {
+        this.annotations = annotations;
+        this.pathComponents = pathComponents;
+        this.callback = callback;
+    }
+
+    int size() {
+        return pathComponents.size();
+    }
+
+    boolean isTerminal(final int pathComponentIndex) {
+        return pathComponentIndex == size();
+    }
+
+    boolean partialMatchAt(final MatchContext context) {
+        int pathComponentIndex = context.getPathComponentIndex();
+
+        if (pathComponentIndex == 0) {
+            return annotationsMatch(context);
+        } else if (pathComponentIndex <= pathComponents.size()) {
+            return pathComponents.get(pathComponentIndex - 1).matches(context);
+        }
+
+        return false;
+    }
+
+    BiFunction<IonReader, T, Integer> getCallback() {
+        return callback;
+    }
+
+    private boolean annotationsMatch(final MatchContext context) {
+        if (annotations.length == 0) {
+            return true;
```

Comment:
This is a bit counter-intuitive to me. All other uses of annotations in a search path require that the matched value have the same annotations in the same order. However, here we treat the absence of annotations as a wildcard instead of as an empty annotation list. How would a user specify that they wanted un-annotated matches?

Reply:
The other case is an annotated `PathComponent`s, e.g. `A::1`. The annotation matching for that case is handled by a decorator and since you only have the decorator when you have annotations the check is different than the SearchPath

That being said it'd be better to normalize those. Thanks for pointing that out